zospy/analyses/new/base.py

```diff
@@ -0,0 +1,560 @@
+
+        return output
+
+    def __getattr__(self, item):
+        return getattr(self._analysis, item)
+
+    def __dir__(self):
+        return sorted(set(super().__dir__()).union(dir(self._analysis)))
+
+
+def new_analysis(
+    oss: OpticStudioSystem, analysis_type: constants.Analysis.AnalysisIDM, settings_first: bool = True
+) -> Analysis:
+    """
+    Creates a new analysis in Zemax.
+
+    Parameters
+    ----------
+    oss: OpticStudioSystem
+        The Zemax OpticStudio system
+    analysis_type: zospy.constants.Analysis.AnalysisIDM
+        Analysis type from `ZOSAPI.Analysis.AnalysisIDM`
+    settings_first: bool
+        Do not run the analysis immediately, which allows to adjust settings. Defaults to `True`.
+
+    Returns
+    -------
+    analysis: Analysis
+
+    Examples
+    --------
+    >>> import zospy as zp
+    >>> new_analysis(oss, zp.constants.Analysis.AnalysisIDM.ZernikeStandardCoefficients)
+    """
+    analysis = (
+        oss.Analyses.New_Analysis_SettingsFirst(analysis_type)
+        if settings_first
+        else oss.Analyses.New_Analysis(analysis_type)
+    )
+    return Analysis(analysis)
+
+
+def _create_tempfile(path: str | None, suffix: str) -> (str, bool):
+    if path is None:
+        fd, path = mkstemp(suffix=suffix, prefix="zospy_")
+        os.close(fd)
+        clean_file = True
+    else:
+        if not path.endswith(suffix):
+            raise ValueError(f"cfgfile should end with '{suffix}'")
+
+        clean_file = False
+
+    return path, clean_file
+
+
+def _handle_complete(self, analysis: Analysis, oncomplete: OnComplete | str) -> None:
+    """Completes the analysis by either closing, releasing or sustaining it.
+
+    Parameters
+    ----------
+    oncomplete: OnComplete | str
+        Action to perform on completion.
+    """
+    oncomplete = OnComplete(oncomplete)
+
+    if oncomplete == OnComplete.Close:
+        analysis.Close()
+    elif oncomplete == OnComplete.Release:
+        analysis.Release()
+    elif oncomplete == OnComplete.Sustain:
+        return
+    else:
+        raise ValueError(f"oncomplete should be a member of zospy.analyses.base.OnComplete, got {oncomplete}")
+
+
+class AnalysisWrapper(ABC, Generic[AnalysisData, AnalysisSettings]):
+    TYPE: str = None
+
+    # Flags to indicate if the analysis needs a configuration file or text output file
+    _needs_config_file: bool = False
+    _needs_text_output_file: bool = False
+
+    def __init__(self, settings: AnalysisSettings, settings_arguments: dict[str, any]):
+        self._init_settings(settings, settings_arguments)
+
+        self._config_file = None
+        self._text_output_file = None
+
+        self._oss = None
+        self._analysis = None
+        self._remove_config_file = False
+        self._remove_text_output_file = False
+
+    def _init_settings(self, settings: AnalysisSettings, parameters: dict[str, any]):
+        parameters.pop("self", None)  # Remove self from parameters if it exists
```

Comment:
In the analyses, we now use:
```python
    def __init__(
        self,
        sampling: str | int = "32x32",
        ...,
        settings: PolarizationTransmissionSettings | None = None,
    ):
        super().__init__(settings or PolarizationTransmissionSettings(), locals())
```

I think `settings` are also in `locals()` in this case. Should we therefore not change this to:
```suggestion
        parameters.pop("self", None)  # Remove self from parameters if it exists
        parameters.pop("settings", None)  # Remove settings from parameters if it exists
```

Reply:
You're right, I added the `settings` parameter after writing the `_init_settings` method, and didn't update the latter afterwards. I made the method more generic now, so instead of popping some variables, it iterates through the fields of the setting class to update them.

Reply:
I like this implementation! We should be wary that we don't implement **kwargs in any analysis class, as then one could specify e.g. `hx=1` rather than `Hx=1` and it would go unnoticed. But since we do not plan on doing so, maybe we can ignore that. If you agree, we can resolve this conversation